Doc/library/email.header.rst

```diff
@@ -203,3 +202,34 @@ The :mod:`email.header` module also provides the following convenient functions.
    :class:`Header` instance.  Optional *maxlinelen*, *header_name*, and
    *continuation_ws* are as in the :class:`Header` constructor.
 
+
+.. function:: decode_header(header)
+
+   Decode a message header value without converting the character set. The header
+   value is in *header*.
+
+   For historical reasons, this function may return either:
+
+   1. A list of pairs containing each of the decoded parts of the header,
+      ``(decoded_bytes, charset)``, where *decoded_bytes* is always an instance of
+      :class:`bytes`, and *charset* is either:
+         - A lower case string containing the name of the character set specified.
+         - ``None`` for non-encoded parts of the header.
+   2. A list of length 1 containing a pair ``(string, None)``, where
+      *string* is always an instance of :class:`str`.
+
+   Here are examples:
+
+      >>> from email.header import decode_header
+      >>> decode_header('=?iso-8859-1?q?p=F6stal?=')
+      [(b'p\xf6stal', 'iso-8859-1')]
+      >>> decode_header('unencoded_string')
+      [('unencoded_string', None)]
+      >>> decode_header('bar =?utf-8?B?ZsOzbw==?=')
+      [(b'bar ', None), (b'f\xc3\xb3o', 'utf-8')]
+
+   .. note::
+
+      This function exists for for backwards compatibility only. For
+      new code we recommend using :mod:`email.header.decode_header_to_string`.
```

Comment:
How about adding a link to the non-legacy API, or an example using that newer API?

Reply:
Do you mean the non-legacy API as described in https://docs.python.org/3/library/email.html, e.g. `email.parser`?

To my knowledge, there is not any function/method in that API which can be straightforwardly used instead of `email.header.decode_header`.

Reply:
>>> from email.headerregistry import HeaderRegistry
>>> decoder = HeaderRegistry()
>>> decoder('To', '=?utf-8?q?M=C3=A4x?= <foo@bar.com>')
'Mäx <foo@bar.com>'
>>> decoder('To', '=?utf-8?q?M=C3=A4x?= <foo@bar.com>').addresses
(Address(display_name='Mäx', username='foo', domain='bar.com'),)

You really don't want to use the legacy decode_header.  It has many bugs that the new API fixes.

Reply:
Obviously this needs to be better documented...